core/iwasm/compilation/aot_emit_gc.c

```diff
@@ -1456,6 +1457,156 @@ aot_compile_op_array_set(AOTCompContext *comp_ctx, AOTFuncContext *func_ctx,
+{
+    LLVMValueRef len, array_obj, fill_value = NULL, offset, array_len, cmp[2],
+                                 boundary, loop_counter;
+    LLVMBasicBlockRef check_obj_succ, len_gt_zero, len_le_zero, inner_else;
+    LLVMBasicBlockRef fill_loop_header, fill_loop_body;
+    WASMArrayType *compile_time_array_type =
+        (WASMArrayType *)comp_ctx->comp_data->types[type_index];
+    uint8 array_elem_type = compile_time_array_type->elem_type;
+    int i;
+
+    POP_I32(len);
+    /* Get LLVM type based on array_elem_type */
+    if (wasm_is_type_reftype(array_elem_type)) {
+        POP_GC_REF(fill_value);
+    }
+    else if (array_elem_type == VALUE_TYPE_I32
+             || array_elem_type == PACKED_TYPE_I8
+             || array_elem_type == PACKED_TYPE_I16) {
+        POP_I32(fill_value);
+    }
+    else if (array_elem_type == VALUE_TYPE_I64) {
+        POP_I64(fill_value);
+    }
+    else if (array_elem_type == VALUE_TYPE_F32) {
+        POP_F32(fill_value);
+    }
+    else if (array_elem_type == VALUE_TYPE_F64) {
+        POP_F64(fill_value);
+    }
+    else {
+        bh_assert(0);
+    }
+
+    POP_I32(offset);
+    POP_GC_REF(array_obj);
+
+    ADD_BASIC_BLOCK(check_obj_succ, "check array objs succ");
+    MOVE_BLOCK_AFTER_CURR(check_obj_succ);
+
+    BUILD_ISNULL(array_obj, cmp[0], "cmp_obj");
+
+    if (!aot_emit_exception(comp_ctx, func_ctx, EXCE_NULL_ARRAY_OBJ, true,
+                            cmp[0], check_obj_succ))
+        goto fail;
+
+    /* Create if block */
+    ADD_BASIC_BLOCK(len_gt_zero, "len_gt_zero");
+    MOVE_BLOCK_AFTER_CURR(len_gt_zero);
+
+    /* Create else(end) block */
+    ADD_BASIC_BLOCK(len_le_zero, "len_le_zero");
+    MOVE_BLOCK_AFTER(len_le_zero, len_gt_zero);
+
+    /* Create fill_loop_header block */
+    ADD_BASIC_BLOCK(fill_loop_header, "fill_loop_header");
+    MOVE_BLOCK_AFTER(fill_loop_header, len_gt_zero);
+
+    /* Create fill_loop_body block */
+    ADD_BASIC_BLOCK(fill_loop_body, "fill_loop_body");
+    MOVE_BLOCK_AFTER(fill_loop_body, len_gt_zero);
+
+    /* Create inner else block */
+    ADD_BASIC_BLOCK(inner_else, "inner_else");
+    MOVE_BLOCK_AFTER(inner_else, len_gt_zero);
+
+    BUILD_ICMP(LLVMIntSGT, len, I32_ZERO, cmp[0], "cmp_len");
+    BUILD_COND_BR(cmp[0], len_gt_zero, len_le_zero);
+
+    /* Move builder to len > 0 block */
+    SET_BUILDER_POS(len_gt_zero);
+    /* dst_offset > UINT32_MAX - len */
+    if (!(boundary = LLVMBuildAdd(comp_ctx->builder, offset, len, ""))) {
+        aot_set_last_error("llvm build failed.");
+        goto fail;
+    }
+    BUILD_ICMP(LLVMIntUGT, boundary, I32_CONST(UINT32_MAX), cmp[0],
+               "boundary_check1");
+    /* dst_offset + len > wasm_array_obj_length(dst_obj) */
+    if (!aot_array_obj_length(comp_ctx, array_obj, &array_len))
+        goto fail;
+    BUILD_ICMP(LLVMIntUGT, boundary, array_len, cmp[1], "boundary_check2");
+
+    /* logical or above 4 boundary checks */
+    for (i = 1; i < 2; ++i) {
```

Comment:
Seems we can remove this line, and change below cmp[i] to cmp[1]

Reply:
Done